File: binding/src/libqfontdatabase.cpp
#include <QByteArray>
#include <QFont>
#include <QFontDatabase>
#include <QFontInfo>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qfontdatabase.h>
#include "libqfontdatabase.h"
#include "libqfontdatabase.hxx"

QFontDatabase* QFontDatabase_new(const QFontDatabase* other) {
	 return new QFontDatabase(*other);
}

QFontDatabase* QFontDatabase_new2(QFontDatabase* other) {
	 return new QFontDatabase(*other);
}

QFontDatabase* QFontDatabase_new3() {
	 return new QFontDatabase();
}

QFontDatabase* QFontDatabase_new4(const QFontDatabase* param1) {
	 return new QFontDatabase(*param1);
}

void QFontDatabase_CopyAssign(QFontDatabase* self, QFontDatabase* other) {
    *self = *other;
}

void QFontDatabase_MoveAssign(QFontDatabase* self, QFontDatabase* other) {
    *self = std::move(*other);
}

libqt_list QFontDatabase_StandardSizes() {
	return QFontDatabase::standardSizes();
}

libqt_list QFontDatabase_WritingSystems() {
	return QFontDatabase::writingSystems();
}

libqt_list QFontDatabase_WritingSystems2(const libqt_string family) {
	return QFontDatabase::writingSystems(QString::fromUtf8(family.data, family.len));
}

libqt_list QFontDatabase_Families() {
	return QFontDatabase::families();
}

libqt_list QFontDatabase_Styles(const libqt_string family) {
	return QFontDatabase::styles(QString::fromUtf8(family.data, family.len));
}

libqt_list QFontDatabase_PointSizes(const libqt_string family) {
	return QFontDatabase::pointSizes(QString::fromUtf8(family.data, family.len));
}

libqt_list QFontDatabase_SmoothSizes(const libqt_string family, const libqt_string style) {
	return QFontDatabase::smoothSizes(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

libqt_string QFontDatabase_StyleString(const QFont* font) {
	QString _ret = QFontDatabase::styleString(*font);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFontDatabase_StyleString2(const QFontInfo* fontInfo) {
	QString _ret = QFontDatabase::styleString(*fontInfo);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QFont* QFontDatabase_Font(const libqt_string family, const libqt_string style, int pointSize) {
	return new QFont(QFontDatabase::font(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len), pointSize));
}

bool QFontDatabase_IsBitmapScalable(const libqt_string family) {
	return QFontDatabase::isBitmapScalable(QString::fromUtf8(family.data, family.len));
}

bool QFontDatabase_IsSmoothlyScalable(const libqt_string family) {
	return QFontDatabase::isSmoothlyScalable(QString::fromUtf8(family.data, family.len));
}

bool QFontDatabase_IsScalable(const libqt_string family) {
	return QFontDatabase::isScalable(QString::fromUtf8(family.data, family.len));
}

bool QFontDatabase_IsFixedPitch(const libqt_string family) {
	return QFontDatabase::isFixedPitch(QString::fromUtf8(family.data, family.len));
}

bool QFontDatabase_Italic(const libqt_string family, const libqt_string style) {
	return QFontDatabase::italic(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_Bold(const libqt_string family, const libqt_string style) {
	return QFontDatabase::bold(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

int QFontDatabase_Weight(const libqt_string family, const libqt_string style) {
	return QFontDatabase::weight(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_HasFamily(const libqt_string family) {
	return QFontDatabase::hasFamily(QString::fromUtf8(family.data, family.len));
}

bool QFontDatabase_IsPrivateFamily(const libqt_string family) {
	return QFontDatabase::isPrivateFamily(QString::fromUtf8(family.data, family.len));
}

libqt_string QFontDatabase_WritingSystemName(int writingSystem) {
	QString _ret = QFontDatabase::writingSystemName(static_cast<QFontDatabase::WritingSystem>(writingSystem));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QFontDatabase_WritingSystemSample(int writingSystem) {
	QString _ret = QFontDatabase::writingSystemSample(static_cast<QFontDatabase::WritingSystem>(writingSystem));
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QFontDatabase_AddApplicationFont(const libqt_string fileName) {
	return QFontDatabase::addApplicationFont(QString::fromUtf8(fileName.data, fileName.len));
}

int QFontDatabase_AddApplicationFontFromData(const libqt_string fontData) {
	return QFontDatabase::addApplicationFontFromData(QByteArray(fontData.data, fontData.len));
}

libqt_list QFontDatabase_ApplicationFontFamilies(int id) {
	return QFontDatabase::applicationFontFamilies(id);
}

bool QFontDatabase_RemoveApplicationFont(int id) {
	return QFontDatabase::removeApplicationFont(id);
}

bool QFontDatabase_RemoveAllApplicationFonts() {
	return QFontDatabase::removeAllApplicationFonts();
}

void QFontDatabase_AddApplicationFallbackFontFamily(int script, const libqt_string familyName) {
	QFontDatabase::addApplicationFallbackFontFamily(static_cast<QChar::Script>(script), QString::fromUtf8(familyName.data, familyName.len));
}

bool QFontDatabase_RemoveApplicationFallbackFontFamily(int script, const libqt_string familyName) {
	return QFontDatabase::removeApplicationFallbackFontFamily(static_cast<QChar::Script>(script), QString::fromUtf8(familyName.data, familyName.len));
}

void QFontDatabase_SetApplicationFallbackFontFamilies(int param1, const libqt_list familyNames) {
	QFontDatabase::setApplicationFallbackFontFamilies(static_cast<QChar::Script>(param1), *familyNames);
}

libqt_list QFontDatabase_ApplicationFallbackFontFamilies(int script) {
	return QFontDatabase::applicationFallbackFontFamilies(static_cast<QChar::Script>(script));
}

QFont* QFontDatabase_SystemFont(int typeVal) {
	return new QFont(QFontDatabase::systemFont(static_cast<QFontDatabase::SystemFont>(typeVal)));
}

void QFontDatabase_OperatorAssign(QFontDatabase* self, const QFontDatabase* param1) {
	self->operator=(*param1);
}

libqt_list QFontDatabase_Families1(int writingSystem) {
	return QFontDatabase::families(static_cast<QFontDatabase::WritingSystem>(writingSystem));
}

libqt_list QFontDatabase_PointSizes2(const libqt_string family, const libqt_string style) {
	return QFontDatabase::pointSizes(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_IsBitmapScalable2(const libqt_string family, const libqt_string style) {
	return QFontDatabase::isBitmapScalable(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_IsSmoothlyScalable2(const libqt_string family, const libqt_string style) {
	return QFontDatabase::isSmoothlyScalable(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_IsScalable2(const libqt_string family, const libqt_string style) {
	return QFontDatabase::isScalable(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

bool QFontDatabase_IsFixedPitch2(const libqt_string family, const libqt_string style) {
	return QFontDatabase::isFixedPitch(QString::fromUtf8(family.data, family.len), QString::fromUtf8(style.data, style.len));
}

void QFontDatabase_Delete(QFontDatabase* self) {
    delete self;
}

File: binding/include/libqfontdatabase.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qfontdatabase.html
pub const qfontdatabase = struct {

    /// New constructs a new QFontDatabase object.
    pub fn New(other: ?*anyopaque) QtC.QFontDatabase {
        return qtc.QFontDatabase_new(@ptrCast(other));
    }


    /// New2 constructs a new QFontDatabase object.
    pub fn New2(other: ?*anyopaque) QtC.QFontDatabase {
        return qtc.QFontDatabase_new2(@ptrCast(other));
    }


    /// New3 constructs a new QFontDatabase object.
    pub fn New3() QtC.QFontDatabase {
        return qtc.QFontDatabase_new3();
    }


    /// New4 constructs a new QFontDatabase object.
    pub fn New4(param1: ?*anyopaque) QtC.QFontDatabase {
        return qtc.QFontDatabase_new4(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QFontDatabase_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QFontDatabase_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn StandardSizes() []const u8 {
        return qtc.QFontDatabase_StandardSizes();
    }

    pub fn WritingSystems() []const u8 {
        return qtc.QFontDatabase_WritingSystems();
    }

    pub fn WritingSystems2(family: []const u8) []const u8 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_WritingSystems2(family_str);
    }

    pub fn Families() []const u8 {
        return qtc.QFontDatabase_Families();
    }

    pub fn Styles(family: []const u8) []const u8 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_Styles(family_str);
    }

    pub fn PointSizes(family: []const u8) []const u8 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_PointSizes(family_str);
    }

    pub fn SmoothSizes(family: []const u8, style: []const u8) []const u8 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_SmoothSizes(family_str, style_str);
    }

    pub fn StyleString(font: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontDatabase_StyleString(@ptrCast(font));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontdatabase.StyleString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn StyleString2(fontInfo: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontDatabase_StyleString2(@ptrCast(fontInfo));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontdatabase.StyleString2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Font(family: []const u8, style: []const u8, pointSize: i32) QtC.QFont {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_Font(family_str, style_str, pointSize);
    }

    pub fn IsBitmapScalable(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_IsBitmapScalable(family_str);
    }

    pub fn IsSmoothlyScalable(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_IsSmoothlyScalable(family_str);
    }

    pub fn IsScalable(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_IsScalable(family_str);
    }

    pub fn IsFixedPitch(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_IsFixedPitch(family_str);
    }

    pub fn Italic(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_Italic(family_str, style_str);
    }

    pub fn Bold(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_Bold(family_str, style_str);
    }

    pub fn Weight(family: []const u8, style: []const u8) i32 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_Weight(family_str, style_str);
    }

    pub fn HasFamily(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_HasFamily(family_str);
    }

    pub fn IsPrivateFamily(family: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
return qtc.QFontDatabase_IsPrivateFamily(family_str);
    }

    pub fn WritingSystemName(writingSystem: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontDatabase_WritingSystemName(@intCast(writingSystem));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontdatabase.WritingSystemName: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn WritingSystemSample(writingSystem: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QFontDatabase_WritingSystemSample(@intCast(writingSystem));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qfontdatabase.WritingSystemSample: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn AddApplicationFont(fileName: []const u8) i32 {
        const fileName_str = qtc.libqt_string{
    .len = fileName.len,
    .data = fileName.ptr,
};
return qtc.QFontDatabase_AddApplicationFont(fileName_str);
    }

    pub fn AddApplicationFontFromData(fontData: []u8) i32 {
        const fontData_str = qtc.libqt_string{
    .len = fontData.len,
    .data = fontData.ptr,
};
return qtc.QFontDatabase_AddApplicationFontFromData(fontData_str);
    }

    pub fn ApplicationFontFamilies(id: i32) []const u8 {
        return qtc.QFontDatabase_ApplicationFontFamilies(id);
    }

    pub fn RemoveApplicationFont(id: i32) bool {
        return qtc.QFontDatabase_RemoveApplicationFont(id);
    }

    pub fn RemoveAllApplicationFonts() bool {
        return qtc.QFontDatabase_RemoveAllApplicationFonts();
    }

    pub fn AddApplicationFallbackFontFamily(script: i32, familyName: []const u8) void {
        const familyName_str = qtc.libqt_string{
    .len = familyName.len,
    .data = familyName.ptr,
};
qtc.QFontDatabase_AddApplicationFallbackFontFamily(@intCast(script), familyName_str);
    }

    pub fn RemoveApplicationFallbackFontFamily(script: i32, familyName: []const u8) bool {
        const familyName_str = qtc.libqt_string{
    .len = familyName.len,
    .data = familyName.ptr,
};
return qtc.QFontDatabase_RemoveApplicationFallbackFontFamily(@intCast(script), familyName_str);
    }

    pub fn SetApplicationFallbackFontFamilies(param1: i32, familyNames: []const u8) void {
        qtc.QFontDatabase_SetApplicationFallbackFontFamilies(@intCast(param1), @ptrCast(familyNames));
    }

    pub fn ApplicationFallbackFontFamilies(script: i32) []const u8 {
        return qtc.QFontDatabase_ApplicationFallbackFontFamilies(@intCast(script));
    }

    pub fn SystemFont(typeVal: i32) QtC.QFont {
        return qtc.QFontDatabase_SystemFont(@intCast(typeVal));
    }

    pub fn OperatorAssign(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QFontDatabase_OperatorAssign(@ptrCast(self), @ptrCast(param1));
    }

    pub fn Families1(writingSystem: i32) []const u8 {
        return qtc.QFontDatabase_Families1(@intCast(writingSystem));
    }

    pub fn PointSizes2(family: []const u8, style: []const u8) []const u8 {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_PointSizes2(family_str, style_str);
    }

    pub fn IsBitmapScalable2(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_IsBitmapScalable2(family_str, style_str);
    }

    pub fn IsSmoothlyScalable2(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_IsSmoothlyScalable2(family_str, style_str);
    }

    pub fn IsScalable2(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_IsScalable2(family_str, style_str);
    }

    pub fn IsFixedPitch2(family: []const u8, style: []const u8) bool {
        const family_str = qtc.libqt_string{
    .len = family.len,
    .data = family.ptr,
};
const style_str = qtc.libqt_string{
    .len = style.len,
    .data = style.ptr,
};
return qtc.QFontDatabase_IsFixedPitch2(family_str, style_str);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QFontDatabase_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/fontdatabase.html#types
pub const enums = struct {
    pub const WritingSystem = enum {
        pub const Any: i32 = 0;
        pub const Latin: i32 = 1;
        pub const Greek: i32 = 2;
        pub const Cyrillic: i32 = 3;
        pub const Armenian: i32 = 4;
        pub const Hebrew: i32 = 5;
        pub const Arabic: i32 = 6;
        pub const Syriac: i32 = 7;
        pub const Thaana: i32 = 8;
        pub const Devanagari: i32 = 9;
        pub const Bengali: i32 = 10;
        pub const Gurmukhi: i32 = 11;
        pub const Gujarati: i32 = 12;
        pub const Oriya: i32 = 13;
        pub const Tamil: i32 = 14;
        pub const Telugu: i32 = 15;
        pub const Kannada: i32 = 16;
        pub const Malayalam: i32 = 17;
        pub const Sinhala: i32 = 18;
        pub const Thai: i32 = 19;
        pub const Lao: i32 = 20;
        pub const Tibetan: i32 = 21;
        pub const Myanmar: i32 = 22;
        pub const Georgian: i32 = 23;
        pub const Khmer: i32 = 24;
        pub const SimplifiedChinese: i32 = 25;
        pub const TraditionalChinese: i32 = 26;
        pub const Japanese: i32 = 27;
        pub const Korean: i32 = 28;
        pub const Vietnamese: i32 = 29;
        pub const Symbol: i32 = 30;
        pub const Other: i32 = 30;
        pub const Ogham: i32 = 31;
        pub const Runic: i32 = 32;
        pub const Nko: i32 = 33;
        pub const WritingSystemsCount: i32 = 34;
    };

    pub const SystemFont = enum {
        pub const GeneralFont: i32 = 0;
        pub const FixedFont: i32 = 1;
        pub const TitleFont: i32 = 2;
        pub const SmallestReadableFont: i32 = 3;
    };

};
